### ISSUE TITLE
Assigning Nil to a scalar declared with captured type changes it's container metaclass to GenericHOW

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#2169]
## The Problem

In the following code:

     sub type_test (::T $type) {
         my Int $i;
         my T $t;

         say "I1:", $i.WHAT;
         $i = Nil;
         say "I2:", $i.WHAT;

         say $t.HOW;
         say "T1:", $t.WHAT;
         $t = Nil;
         say $t.HOW;
         say "T2:", $t.WHAT;
     }

     type_test( Int );

The last `say` generates the exception (generated with `--ll-exception`):

```
No such method 'gist' for invocant of type 'T'
   at SETTING::src/core/Exception.pm6:57  (/Users/vrurg/perl6/share/perl6/runtime/CORE.setting.moarvm:throw)
 from SETTING::src/core/Exception.pm6:2711  (/Users/vrurg/perl6/share/perl6/runtime/CORE.setting.moarvm:)
 from gen/moar/BOOTSTRAP.nqp:3407  (/Users/vrurg/perl6/share/nqp/lib/Perl6/BOOTSTRAP.moarvm:)
 from SETTING::src/core/io_operators.pm6:19  (/Users/vrurg/perl6/share/perl6/runtime/CORE.setting.moarvm:say)
 from SETTING::src/core/io_operators.pm6:10  (/Users/vrurg/perl6/share/perl6/runtime/CORE.setting.moarvm:say)
 from ./proxy.p6:15  (<ephemeral file>:type_test)
 from ./proxy.p6:18  (<ephemeral file>:<unit>)
 from ./proxy.p6:1  (<ephemeral file>:<unit-outer>)
 from gen/moar/stage2/NQPHLL.nqp:1566  (/Users/vrurg/perl6/share/nqp/lib/NQPHLL.moarvm:eval)
 from gen/moar/stage2/NQPHLL.nqp:1805  (/Users/vrurg/perl6/share/nqp/lib/NQPHLL.moarvm:evalfiles)
 from gen/moar/stage2/NQPHLL.nqp:1728  (/Users/vrurg/perl6/share/nqp/lib/NQPHLL.moarvm:command_eval)
 from src/Perl6/Compiler.nqp:42  (/Users/vrurg/perl6/share/nqp/lib/Perl6/Compiler.moarvm:command_eval)
 from gen/moar/stage2/NQPHLL.nqp:1654  (/Users/vrurg/perl6/share/nqp/lib/NQPHLL.moarvm:command_line)
 from gen/moar/main.nqp:47  (/Users/vrurg/perl6/share/perl6/runtime/perl6.moarvm:MAIN)
 from gen/moar/main.nqp:38  (/Users/vrurg/perl6/share/perl6/runtime/perl6.moarvm:<mainline>)
 from <unknown>:1  (/Users/vrurg/perl6/share/perl6/runtime/perl6.moarvm:<main>)
 from <unknown>:1  (/Users/vrurg/perl6/share/perl6/runtime/perl6.moarvm:<entry>)
```

## Expected Behavior

Must output `T2:(Int)` as it does for `$i`. 

## Actual Behavior

```
I1:(Int)
I2:(Int)
Perl6::Metamodel::ClassHOW.new
T1:(Int)
Perl6::Metamodel::GenericHOW.new
No such method 'gist' for invocant of type 'T'
```

Note the change of ClassHOW to GenericHOW.

## Environment
* Operating system: macOS 10.13
* Compiler version (`perl6 -v`): This is Rakudo version 2018.06 built on MoarVM version 2018.06

**Comment by vrurg** [comment]
Fixed with #2731, tests are pending merge in perl6/roast#543

**Comment by patzim** [comment]
Tests have been merged. Closing!
